python/dgl/sparse/elementwise_op.py

```diff
@@ -2,43 +2,235 @@
+    # returned.
+    # So this also handles the case of scalar * SparseMatrix since we set
+    # SparseMatrix.__rmul__ to be the same as SparseMatrix.__mul__.
+    return NotImplemented
+
+
+def sp_div(A: SparseMatrix, B: Union[SparseMatrix, Scalar]) -> SparseMatrix:
+    """Elementwise division
+
+    If :attr:`B` is a sparse matrix, both :attr:`A` and :attr:`B` must be
+    diagonal matrices.
+
+    Parameters
+    ----------
+    A : SparseMatrix
+        First operand
+    B : SparseMatrix or Scalar
+        Second operand
+
+    Returns
+    -------
+    SparseMatrix
+        Result of A / B
+
+    Examples
+    --------
+    >>> indices = torch.tensor([[1, 0, 2], [0, 3, 2]])
+    >>> val = torch.tensor([1, 2, 3])
+    >>> A = dglsp.spmatrix(indices, val, shape=(3, 4))
+    >>> A / 2
+    SparseMatrix(indices=tensor([[1, 0, 2],
+                                 [0, 3, 2]]),
+                 values=tensor([0.5000, 1.0000, 1.5000]),
+                 shape=(3, 4), nnz=3)
+    """
+    if is_scalar(B):
+        return val_like(A, A.val / B)
+    if A.is_diag() and B.is_diag():
+        assert A.shape == B.shape, (
+            f"The shape of diagonal matrix A {A.shape} and B {B.shape} must"
+            f"match for elementwise division."
+        )
+        return diag(A.val / B.val, A.shape)
+    # Python falls back to B.__rtruediv__(A) then TypeError when NotImplemented
+    # is returned.
+    return NotImplemented
+
+
+def sp_power(A: SparseMatrix, scalar: Scalar) -> SparseMatrix:
+    """Take the power of each nonzero element and return a sparse matrix with
+    the result.
+
+    Parameters
+    ----------
+    A : SparseMatrix
+        Sparse matrix
+    scalar : float or int
+        Exponent
+
+    Returns
+    -------
+    SparseMatrix
+        Sparse matrix
+
+    Examples
+    --------
+    >>> indices = torch.tensor([[1, 0, 2], [0, 3, 2]])
+    >>> val = torch.tensor([10, 20, 30])
+    >>> A = dglsp.spmatrix(indices, val)
+    >>> A ** 2
+    SparseMatrix(indices=tensor([[1, 0, 2],
+                                 [0, 3, 2]]),
+    values=tensor([100, 400, 900]),
+    shape=(3, 4), nnz=3)
+    """
+    # Python falls back to scalar.__rpow__ then TypeError when NotImplemented
+    # is returned.
+    return val_like(A, A.val**scalar) if is_scalar(scalar) else NotImplemented
 
-    The supported combinations are shown as follows.
 
-    +--------------+------------+--------------+--------+
-    |    A \\ B    | DiagMatrix | SparseMatrix | scalar |
-    +--------------+------------+--------------+--------+
-    |  DiagMatrix  |     ✅     |      ✅      |   🚫   |
-    +--------------+------------+--------------+--------+
-    | SparseMatrix |     ✅     |      ✅      |   🚫   |
-    +--------------+------------+--------------+--------+
-    |    scalar    |     🚫     |      🚫      |   🚫   |
-    +--------------+------------+--------------+--------+
+SparseMatrix.__add__ = sp_add
```

Comment:
We do we need to duplicate each method?
What's the downside of SparseMatrix.__add__ = add?

In general the PR is oversize, in which 2 things are done.
One remove Python DiagMatrix, and use a new mechanism from the ops.
Consider breaking them to 2 PRs for easier to review.

Reply:
> We do we need to duplicate each method?
> What's the downside of SparseMatrix.add = add?

I didn't get your point. Do you mean why we don't define SparseMatrix.__add__ in sparse_matrix.py?

> Consider breaking them to 2 PRs for easier to review.

Done.

Reply:
The first comment is not related to this PR anymore.